elfcore/src/coredump.rs

```diff
@@ -1328,28 +1337,53 @@ fn write_va_region<T: Write>(
     va_region: &VaRegion,
     pv: &ProcessView,
 ) -> Result<usize, CoreError> {
+    // For optimal performance should be in [8KiB; 64KiB] range.
+    // Selected 64 KiB as data on various hardware platforms shows
+    // peak performance in this case.
+    const BUFFER_SIZE: usize = 0x10000;
+
     let mut dumped = 0_usize;
     let mut address = va_region.begin;
+    let mut buffer = [0_u8; BUFFER_SIZE];
 
     while address < va_region.end {
-        let mut page = vec![0xF1_u8; pv.page_size];
-
-        // SAFETY: Reading other process memory. The process is paused,
-        // the address comes from the kernel. The structure is repr(C) with no padding bytes,
-        // so all byte patterns are valid.
-        let _bytes_read = unsafe {
-            nix::libc::pread64(
-                pv.mem.as_raw_fd(),
-                page.as_mut_ptr() as *const _ as *mut c_void,
-                page.len(),
-                address as i64,
-            )
-        };
-
-        writer.write_all(page.as_bytes())?;
-
-        address += pv.page_size as u64;
-        dumped += page.as_bytes().len();
+        let len = std::cmp::min((va_region.end - address) as usize, BUFFER_SIZE);
+        match process_vm_readv(
+            pv.pid,
+            &[IoVec::from_mut_slice(&mut buffer)],
+            &[RemoteIoVec {
+                base: address as usize,
+                len,
+            }],
+        ) {
+            Ok(bytes_read) => {
+                writer.write_all(&buffer[..bytes_read])?;
+
+                address += bytes_read as u64;
+                dumped += bytes_read;
+            }
+            Err(_) => {
+                // Every precaution has been taken to read the accessible
+                // memory only and still something has gone wrong. Nevertheless,
+                // have to make forward progress to dump exactly as much memory
+                // as the caller expects.
+                //
+                // Save dummy data up to the next page boundary.
+
+                // Page size is a power of 2 on modern platforms, that is asserted in
+                // the `ProcessView`'s constructor. Round up with bit twiddling.
```

Comment:
could throw a debug_assert! for page size in here if you wanted

Reply:
Thank, great idea! Now the assert is right at the place where the assumption is made, very nice.